<a href="https://colab.research.google.com/github/nihalhabeeb/hotel-booking-analysis/blob/main/Hotel_Booking_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

Mounting google drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


The first step is to import neccessary libraries

In [ ]:
## Importing Libraries
import pandas as pd
import numpy as np
import seaborn as sbrn
import matplotlib.pyplot as plt

Download the dataset to use it as a pandas dataframe 

In [ ]:
file_path = '/content/drive/MyDrive/AlmaBetter/Capstone Projects/EDA/Hotel Bookings.csv'
hotel_df = pd.read_csv(file_path)  

Lets find the shape of the dataset.

In [ ]:
hotel_df.shape

(119390, 32)

Great..! We know that our dataset has 119390 rows and 32 columns

In [ ]:
list(hotel_df.columns)

['hotel',
 'is_canceled',
 'lead_time',
 'arrival_date_year',
 'arrival_date_month',
 'arrival_date_week_number',
 'arrival_date_day_of_month',
 'stays_in_weekend_nights',
 'stays_in_week_nights',
 'adults',
 'children',
 'babies',
 'meal',
 'country',
 'market_segment',
 'distribution_channel',
 'is_repeated_guest',
 'previous_cancellations',
 'previous_bookings_not_canceled',
 'reserved_room_type',
 'assigned_room_type',
 'booking_changes',
 'deposit_type',
 'agent',
 'company',
 'days_in_waiting_list',
 'customer_type',
 'adr',
 'required_car_parking_spaces',
 'total_of_special_requests',
 'reservation_status',
 'reservation_status_date']

describe() method can be used to get statistical information about the dataset. We were able to make some very useful findings using this information (which comes later in the notebook).

In [ ]:
hotel_df.describe()

is_canceled  ...  total_of_special_requests
count  119390.000000  ...              119390.000000
mean        0.370416  ...                   0.571363
std         0.482918  ...                   0.792798
min         0.000000  ...                   0.000000
25%         0.000000  ...                   0.000000
50%         0.000000  ...                   0.000000
75%         1.000000  ...                   1.000000
max         1.000000  ...                   5.000000

[8 rows x 20 columns]

In [ ]:
hotel_df.dtypes

hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent                             

In [ ]:
hotel_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

## Preparing and Cleaning the Dataset

### Handing Null Values

Now lets find out how many cells are missing from our data set

In [ ]:
hotel_df.isnull().sum().sort_values(ascending = False)

company                           112593
agent                              16340
country                              488
children                               4
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
is_canceled                            0
market_segment                         0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
babies                                 0
meal                                   0
reservation_status_date                0
distribution_channel                   0
reservation_status                     0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes 

The percentage of null values in each columns give a better picture and will help us to decide if any columns have to be removed.

In [ ]:
hotel_df.isnull().sum().sort_values(ascending = False)/len(hotel_df.index)*100

company                           94.306893
agent                             13.686238
country                            0.408744
children                           0.003350
lead_time                          0.000000
arrival_date_year                  0.000000
arrival_date_month                 0.000000
arrival_date_week_number           0.000000
is_canceled                        0.000000
market_segment                     0.000000
arrival_date_day_of_month          0.000000
stays_in_weekend_nights            0.000000
stays_in_week_nights               0.000000
adults                             0.000000
babies                             0.000000
meal                               0.000000
reservation_status_date            0.000000
distribution_channel               0.000000
reservation_status                 0.000000
is_repeated_guest                  0.000000
previous_cancellations             0.000000
previous_bookings_not_canceled     0.000000
reserved_room_type              

The 'company' column has aroun 94% null values! The 'agent' column has much less but still non-negligible amount (around 14%) of null values. The 'country' and 'children' have negligible amount of null values.

In [ ]:
hotel_df['agent'][10]

240.0

The 'agent' column consists of float values. A little research about the dataset on the internet revealed that this column refers to the agent's ID. So, the null values cannot be replaced by a value calculated from the rest of the values to make any sense. Since, they are just IDs it is better to remove the column entirely as they won't be necessary in any analysis considering the fact that a good amount of null values are present.

The 'company' column can be removed easily as most of the values are null. Those values also refer to the company ID and removing them is not an issue anyway.

In [ ]:
hotel_df = hotel_df.drop(columns=['company','agent'],axis=1)

The 'children' and 'country' columns have some null values. Those rows which contain the null values can be removed from the analysis.

In [ ]:
# dropping the rows with null values completely
hotel_df = hotel_df.dropna(axis=0)

### Handling Some Data That Don't Make Any Sense!

From the describe() method used earlier we noticed something. The minimum value of adults column is zero. While it is possible that there are bookings done by children only without any adult present it is better to check those data carefully.

In [ ]:
# getting the data with adults column having zero value
hotel_df[hotel_df.adults == 0].head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
2224  Resort Hotel            0  ...           Check-Out               2015-10-06
2409  Resort Hotel            0  ...           Check-Out               2015-10-12
3181  Resort Hotel            0  ...           Check-Out               2015-11-23
3684  Resort Hotel            0  ...           Check-Out               2016-01-04
3708  Resort Hotel            0  ...           Check-Out               2016-01-05

[5 rows x 30 columns]

There are bookings where the number of adults, children and babies are zero! That would imply there are zero guests in the booking which does not make any sense. We decided to consider such data as incorrect and exclude them from the analysis.

A new column 'Total Guests' is created.

In [ ]:
hotel_df['total_guests'] = hotel_df['adults'] + hotel_df['children'] + hotel_df['babies']

We select only those data where all the three columns (adults, babies and children) are zero i.e the total guests are zero.

In [ ]:
# dropping those rows which have zero total guests
hotel_df.drop(hotel_df[hotel_df['total_guests']==0].index, inplace = True)

In [ ]:
# checking if there are any rows with zero total guests
hotel_df[hotel_df.total_guests == 0]

Empty DataFrame
Columns: [hotel, is_canceled, lead_time, arrival_date_year, arrival_date_month, arrival_date_week_number, arrival_date_day_of_month, stays_in_weekend_nights, stays_in_week_nights, adults, children, babies, meal, country, market_segment, distribution_channel, is_repeated_guest, previous_cancellations, previous_bookings_not_canceled, reserved_room_type, assigned_room_type, booking_changes, deposit_type, days_in_waiting_list, customer_type, adr, required_car_parking_spaces, total_of_special_requests, reservation_status, reservation_status_date, total_guests]
Index: []